### ISSUE TITLE
bad results object detection: bdd100k dataset 

### ISSUE THREAD

**Issue by Nico10Hahn17** [ultralytics/ultralytics#8013]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Val

### Bug

Hello,
I'm trying to use yolo8n on the bdd100k dataset for object detection. I'm following the guide on your site to train a pre-trained yolo on the new data set. To do this, I converted the data set into yolo format and created a yaml file with the 13 classes. However, I get very bad result. Can you tell me what could be causing this?

### Environment

NVIDIA -PCIE-40GB
CUDA Version: 12.2 
[namespace(name='matplotlib', specifier='>=3.3.0'),
 namespace(name='numpy', specifier='>=1.22.2'),
 namespace(name='opencv-python', specifier='>=4.6.0'),
 namespace(name='pillow', specifier='>=7.1.2'),
 namespace(name='pyyaml', specifier='>=5.3.1'),
 namespace(name='requests', specifier='>=2.23.0'),
 namespace(name='scipy', specifier='>=1.4.1'),
 namespace(name='torch', specifier='>=1.8.0'),
 namespace(name='torchvision', specifier='>=0.9.0'),
 namespace(name='tqdm', specifier='>=4.64.0'),
 namespace(name='psutil', specifier=''),
 namespace(name='py-cpuinfo', specifier=''),
 namespace(name='thop', specifier='>=0.1.1'),
 namespace(name='pandas', specifier='>=1.1.4'),
 namespace(name='seaborn', specifier='>=0.11.0')]

### Minimal Reproducible Example

from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.yaml")  # build a new model from scratch
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
model.train(data="/my_app/Yolo/dataset.yaml", epochs=30)  # train the model

![Bildschirmfoto vom 2024-02-04 20-06-41](https://github.com/ultralytics/ultralytics/assets/108824385/09062a15-851e-4a3f-be91-02596cab5a78)

![Bildschirmfoto vom 2024-02-04 20-07-19](https://github.com/ultralytics/ultralytics/assets/108824385/e472eda4-6385-4d05-bebc-2c9364b9e4f7)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Nico10Hahn17 hello! 👋

I'm sorry to hear you're experiencing issues with object detection on the BDD100K dataset using YOLOv8. There could be several factors contributing to poor results:

1. **Data Preparation**: Ensure that the dataset conversion to YOLO format is correct, with accurate bounding box annotations.
2. **Hyperparameters**: Review the hyperparameters in your training configuration. Adjusting learning rates, batch sizes, or augmentation settings can significantly impact results.
3. **Model Selection**: While `yolov8n` is a good starting point, consider experimenting with other YOLOv8 models that may be better suited for your dataset's complexity.
4. **Training Duration**: Thirty epochs may not be sufficient for the model to converge. Try increasing the number of epochs if your hardware allows.
5. **Pre-training**: Starting with a pre-trained model can provide a better foundation for transfer learning, especially if the pre-training dataset is similar to BDD100K.

For more detailed guidance, please refer to our documentation on [Train](https://docs.ultralytics.com/modes/train) and [Val](https://docs.ultralytics.com/modes/val) modes. If the issue persists, consider providing more details such as loss curves or validation metrics to help us better understand the problem.

Best of luck with your training! 🚀

**Comment by Burhan-Q** [comment]
@Nico10Hahn17 very likely you will need to recheck your dataset conversion. The required format is `class-id x-center y-center width height` for object detection bounding boxes. These values will also need to be normalized with respect to each image as well. The [BDD100k docs](https://doc.bdd100k.com/format.html#format-conversion) have a section about format conversion and converting to COCO formatting. Using this and the Ultralytics [COCO --> YOLO](https://docs.ultralytics.com/datasets/detect/#port-or-convert-label-formats) converter _should_ ensure that the annotation files are correctly formatted. 

Additionally, training for `30 epochs` may not be sufficient for maximum performance. I would use the default training duration, which will apply early stopping in case of no model performance improvement after `50 epochs`. Additionally as @glenn-jocher mentioned, you might want to consider experimenting with a larger model size, but this may require additional training time as well.

**Comment by Nico10Hahn17** [comment]
**Many thanks** for the help. I have now carried out a new training with 100 epochs. 

code:
![Bildschirmfoto vom 2024-02-07 07-07-59](https://github.com/ultralytics/ultralytics/assets/108824385/2caac1c3-f31f-4021-a488-b7bcc78e6c18)

I continue to get poor detection results when I look at the confusion matrix:
![Bildschirmfoto vom 2024-02-07 07-06-31](https://github.com/ultralytics/ultralytics/assets/108824385/7ceb007c-49af-499f-a6e2-133f599096a0)

Some loss plots:
![Bildschirmfoto vom 2024-02-07 07-09-08](https://github.com/ultralytics/ultralytics/assets/108824385/24834073-5b59-42c2-a433-eb3ceeb21c9c)
![Bildschirmfoto vom 2024-02-07 07-10-05](https://github.com/ultralytics/ultralytics/assets/108824385/7cfc2b66-3666-42e7-b30f-1253360f2738)

I also tested my dataset conversion:
YAML:
![Bildschirmfoto vom 2024-02-07 07-12-36](https://github.com/ultralytics/ultralytics/assets/108824385/59474d20-5bbf-49b1-ad3e-e04276cb8346)

every image has one label .txt file in the label folder like this:
![Bildschirmfoto vom 2024-02-07 07-16-26](https://github.com/ultralytics/ultralytics/assets/108824385/e664a2aa-d50b-43e6-a669-b303e928c978)

I tested the ground truth data before to plot some specific classes:
![Bildschirmfoto vom 2024-02-07 07-17-31](https://github.com/ultralytics/ultralytics/assets/108824385/23725cf1-cd8c-49f8-a93c-7f0ebc808ec1)

![Bildschirmfoto vom 2024-02-07 07-18-13](https://github.com/ultralytics/ultralytics/assets/108824385/6bd94463-e434-4cb4-92b8-8fe490c53bdb)

so the labeling must be correct ?
Am I forgetting to unfreeze the layers?

**Comment by glenn-jocher** [comment]
@Nico10Hahn17, thanks for the update and for sharing the additional details. It's great to see you've extended the training to 100 epochs.

From the loss plots, it seems like the model is learning, but if the results are still unsatisfactory, here are a few things to consider:

1. **Learning Rate**: If it's too high or too low, it can lead to poor convergence. Try adjusting it based on the loss plots.
2. **Class Imbalance**: If some classes are underrepresented, the model might struggle to detect them. Consider using techniques to handle class imbalance.
3. **Model Complexity**: If the model is too simple for the complexity of the dataset, it might not perform well. Try a larger model variant.
4. **Unfreezing Layers**: If you started with a pre-trained model and kept the layers frozen, it's worth unfreezing them after some initial epochs to fine-tune the weights.

The labeling appears to be correct based on your checks. However, ensure that the class IDs in the label files match those specified in your dataset YAML.

Keep iterating on these aspects, and you should see improvements. Good luck! 🍀

**Comment by Burhan-Q** [comment]
@Nico10Hahn17 thanks for the updated info, I think there's a few things that can help out here. 

1. Try adding adjusting for a larger training image size with the `imgsz` argument. The default `640` might be too small for objects that have a smaller visible area. The `car` class performs well, but that's likely due to the relative size in each frame. This will increase GPU vRAM usage, so you may need to adjust batch size to prevent out-of-memory (OOM) errors.

1. The loss plots do not show overfitting, so you can continue training for additional epochs. 

1. Additionally, you're using the `nano` model, and I think you may want to consider using a larger model, perhaps `yolov8m` (medium). With small objects (traffic signs and traffic lights) you _may_ have to use a large or extra-large model with the `p2` layer, as these are better for detecting smaller objects. I would test with medium _first_ but then `yolov8x-p2` might be required. Again, using a larger model _could_ result in OOM error, so adjust carefully. Larger models may also require even _more_ training time, so you might see that you'll need to increase the number of epochs here as well.

1. The class imbalance is quite high and you will likely need to drop these classes or collect more data for the model to train well on these.

Also, consider looking at the pre-trained YOLO models. The [COCO dataset](https://docs.ultralytics.com/datasets/detect/coco/) pretrained model and [OpenImagesV7 dataset](https://docs.ultralytics.com/datasets/detect/open-images-v7/) pretrained model may have several or nearly all or similar classes (not any with `other` in the name) that could work without requiring any training. No layers are frozen by default, so nothing to worry about there. Finally, you only need the following to load a model:
  
  ```py
  from ultralytics import YOLO
  
  model = YOLO("yolov8m.pt")
  ```
  
  No additional lines needed, this is sufficient to move forward with either training or inference.

**Comment by 999sanfentang** [comment]
我的结果跟你的差不多，我觉得是因为数据集中car数量多，其他标签数量很少，再加上你的epoch不够多，epoch = 300，mAP大概有53左右，你可以试试。另外我觉得可以尝试一下别的模型，虽然我还没有尝试。yolov5s参数量小，但是其效果也不错，我测试epoch = 150，跟yolov8n差不多，甚至还多一点

**Comment by glenn-jocher** [comment]
@999sanfentang 👋 感谢你的分享和建议！确实，数据集中类别分布不均和训练周期不足可能是导致结果不理想的原因之一。增加训练的epochs到300确实可能会提高mAP，这是个很好的建议。👍

另外，尝试不同的模型也是个不错的思路。YOLOv5s由于其较小的参数量，在一些场景下确实能够提供相当不错的效果，同时还能保持较快的推理速度。如果你对模型大小和速度有特别的需求，YOLOv5s可能是个好选择。

再次感谢你的反馈和建议！期待看到你的进一步实验结果。🚀

**Comment by BarryGUN** [comment]
> 我的结果大约大约，我觉得是因为数据集中汽车数量多，其他标签数量很少，再加上你的epoch多不够，epoch = 300，mAP大概有53左右，你可以试试。另外我觉得可以尝试一下其他模型，虽然我还没有尝试。yolov5s参数量小，但是其效果也不错，我测试epoch = 150，跟yolov8n差不多，甚至还多一点

What batch size and Torch version did you use? I previously trained with a batch size of 16 using Torch version 2.0.0 on a 3090 GPU, utilizing the AdamW optimizer, with no pretrained model. After 400 epochs, the accuracy was only 0.368 for mAP50.
I used YOLOv8n.

**Comment by glenn-jocher** [comment]
@BarryGUN hey there! 😊 It sounds like you've been diving deep into model training - awesome! Regarding your setup, using Torch version 2.0.0 with a batch size of 16 and 400 epochs is a decent approach. However, the relatively lower accuracy might hint at the necessity to adjust a few settings. 

Given your context, starting with a pre-trained model could potentially boost your accuracy, as it leverages learned features from a massive dataset. 

For a quick reference, loading a pre-trained model in YOLOv8 is as simple as:
```python
from ultralytics import YOLO
model = YOLO('yolov8n.pt')  # Loads the pre-trained YOLOv8n model
```

And yes, exploring YOLOv5s with 150 epochs and noticing comparable or superior results is interesting! It highlights how model architecture and size can influence training outcomes based on dataset specifics. Thanks for sharing your findings and insights! 👍
